bastionzero/bzero
# Lazy User Creation

## Description of the change

This code empowers the agent to create a user if it sees a user trying to connect to one that doesn't exist. This is so that we can support bzero-user and ssm-user without having to create both on every machine and allowing us to elegantly transition later.

Creates a new sudoers file and adds users to that file 
```
# Created by the BastionZero Agent on 2022-06-02 16:38:50 +0000 UTC

ssm-user ALL=(ALL) NOPASSWD:ALL
bzero-user ALL=(ALL) NOPASSWD:ALL
```

## Testing

This functionality should work for both ssh and shell. 
1. Add one (or both) of the following users to your target connect policy: "ssm-user", "bzero-user"
2. Connect as one of those users
3. Neither of those users should exist on the box but you will be able to connect, once you've connected you can verify that they exist and do have sudoer priviledges.

**backend branch:** 
**zli branch:** 

#### Ready to run system tests?


## Relevant release note information

Release Notes: Adds the ability to login as a user that doesn't exist on the machine and will create that user (if it's allowed to).

## Related JIRA tickets

Relates to JIRA: CWC-1901

## Have you considered the security impacts?

Does this PR have any security impact?


If yes, please explain: The one thing that I want to bring attention to, is that when we create a sudoers file in the /etc/sudoers.d folder, we create it with 640 so that we can maintain the ability to write to it. Usually, this file is supposed to have 440 permissions. I don't think this is a very big deal and I will say that only creating users on a as-need basis is the better security move as opposed to adding them all to the sudoers file from the get-go and locking the file down.
